File: week-1/problem-set/prob3.py
"""Prob3 example.

This file belongs to the worked week 1 problem-set solutions. It explores the
classic bridge-and-torch puzzle with a brute-force search over the first two
crossings.
"""

import itertools

times = [1, 2, 5, 10]


def person_label(time_value):
    """Return a readable label for the person with the given crossing time.

    Args:
        time_value (object): The time_value input used by this function.

    Returns:
        str: The string produced by the algorithm.
    """
    person_number = times.index(time_value) + 1
    return f"Person {person_number} ({time_value} min)"


def cross_bridge(combination, arrived, solution):
    """Simulate one crossing and the fastest possible return with the torch.

    Args:
        combination (tuple): The combination input used by this function.
        arrived (list): The arrived input used by this function.
        solution (object): The solution steps built so far.

    Returns:
        tuple: The tuple of values produced by the algorithm.
    """
    crossing_time = max(combination)

    # The chosen pair crosses to the far side.
    updated_arrived = arrived.copy()
    updated_arrived.extend(combination)

    # The fastest person on the far side brings the torch back.
    flash_return = min(updated_arrived)
    updated_arrived.remove(flash_return)

    remaining = [time for time in times if time not in updated_arrived]
    step_text = (
        f"{person_label(combination[0])} and {person_label(combination[1])} cross "
        f"({crossing_time} min). {person_label(flash_return)} returns "
        f"({flash_return} min)."
    )

    updated_solution = solution + [step_text]
    return updated_arrived, remaining, crossing_time + flash_return, updated_solution


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    time_limit = 17
    solutions = []

    for first_pair in itertools.combinations(times, 2):
        arrived_after_first, remaining_after_first, first_cost, first_steps = cross_bridge(
            first_pair, [], []
        )

        for second_pair in itertools.combinations(remaining_after_first, 2):
            arrived_after_second, remaining, second_cost, second_steps = cross_bridge(
                second_pair, arrived_after_first, first_steps
            )

            final_crossing = max(remaining)
            total_time = first_cost + second_cost + final_crossing

            if total_time <= time_limit:
                final_step = (
                    f"{person_label(remaining[0])} and {person_label(remaining[1])} cross "
                    f"({final_crossing} min)."
                )
                solutions.append((total_time, second_steps + [final_step]))

    for solution_number, (total_time, steps) in enumerate(sorted(solutions), start=1):
        print(f"Solution {solution_number}:")
        for step in steps:
            print(f"  {step}")
        print(f"  Total time: {total_time} minutes\n")


if __name__ == "__main__":
    main()

File: week-1/problem-set/prob3.cpp
/**
 * @file prob3.cpp
 * @brief C++ version of the bridge-and-torch puzzle search.
 */

#include <algorithm>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

using namespace std;

struct BridgeState {
    vector<int> arrived;
    vector<int> remaining;
    int travel_time;
    vector<string> steps;
};

/**
 * @brief Return a readable label for the person with the given crossing time.
 * @param times const vector<int>& Ordered list of crossing times. Each value has type int.
 * @param time_value int Crossing time of the person being labelled.
 * @return string Human-readable description such as "Person 1 (1 min)".
 */
string person_label(const vector<int>& times, int time_value) {
    int person_number = static_cast<int>(find(times.begin(), times.end(), time_value) - times.begin()) + 1;
    return "Person " + to_string(person_number) + " (" + to_string(time_value) + " min)";
}

/**
 * @brief Simulate one crossing and the fastest possible torch return.
 * @param combination pair<int, int> Two crossing times chosen for the current move.
 * @param arrived const vector<int>& People already on the far side after earlier moves.
 * @param solution const vector<string>& Solution steps collected so far.
 * @param times const vector<int>& Ordered list of all crossing times.
 * @return BridgeState Updated state after the crossing and the return trip.
 */
BridgeState cross_bridge(pair<int, int> combination,
                         const vector<int>& arrived,
                         const vector<string>& solution,
                         const vector<int>& times) {
    int crossing_time = max(combination.first, combination.second);
    vector<int> updated_arrived = arrived;
    updated_arrived.push_back(combination.first);
    updated_arrived.push_back(combination.second);

    int flash_return = *min_element(updated_arrived.begin(), updated_arrived.end());
    updated_arrived.erase(find(updated_arrived.begin(), updated_arrived.end(), flash_return));

    vector<int> remaining;
    for (int time_value : times) {
        if (find(updated_arrived.begin(), updated_arrived.end(), time_value) == updated_arrived.end()) {
            remaining.push_back(time_value);
        }
    }

    string step_text = person_label(times, combination.first) + " and " +
                       person_label(times, combination.second) + " cross (" +
                       to_string(crossing_time) + " min). " +
                       person_label(times, flash_return) + " returns (" +
                       to_string(flash_return) + " min).";

    vector<string> updated_solution = solution;
    updated_solution.push_back(step_text);

    return {updated_arrived, remaining, crossing_time + flash_return, updated_solution};
}

/**
 * @brief Run the bridge-and-torch search and print every solution within the time limit.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> times = {1, 2, 5, 10};
    int time_limit = 17;
    vector<pair<int, vector<string>>> solutions;

    for (size_t i = 0; i < times.size(); ++i) {
        for (size_t j = i + 1; j < times.size(); ++j) {
            BridgeState first_state = cross_bridge({times[i], times[j]}, {}, {}, times);

            for (size_t a = 0; a < first_state.remaining.size(); ++a) {
                for (size_t b = a + 1; b < first_state.remaining.size(); ++b) {
                    BridgeState second_state = cross_bridge(
                        {first_state.remaining[a], first_state.remaining[b]},
                        first_state.arrived,
                        first_state.steps,
                        times
                    );

                    int final_crossing = max(second_state.remaining[0], second_state.remaining[1]);
                    int total_time = first_state.travel_time + second_state.travel_time + final_crossing;

                    if (total_time <= time_limit) {
                        vector<string> steps = second_state.steps;
                        steps.push_back(
                            person_label(times, second_state.remaining[0]) + " and " +
                            person_label(times, second_state.remaining[1]) + " cross (" +
                            to_string(final_crossing) + " min)."
                        );
                        solutions.push_back({total_time, steps});
                    }
                }
            }
        }
    }

    sort(solutions.begin(), solutions.end(), [](const auto& left, const auto& right) {
        return left.first < right.first;
    });

    for (size_t index = 0; index < solutions.size(); ++index) {
        cout << "Solution " << (index + 1) << ":\n";
        for (const string& step : solutions[index].second) {
            cout << "  " << step << '\n';
        }
        cout << "  Total time: " << solutions[index].first << " minutes\n\n";
    }

    return 0;
}
